sdk/how-to-guides/ports.md

```diff
@@ -0,0 +1,144 @@
+> See first: [Juju `open-port`](https://juju.is/docs/juju/hook-tool#open-port)
+
+**Contents**:
+
+1. [Open a port based on the charm config](#heading--open-a-port-to-your-charm)
+1. [Test the port is opened](#heading--test-the-port-is-opened)
+
+<a href="#heading--heading--open-a-port-to-your-charm"><h2 id="heading--heading--open-a-port-to-your-charm">Open a port based on the charm config</h2></a>
+
+To make a service permanently reachable under a stable URL on the cluster the charm needs to open (expose) a port. This ensures that the charm will be consistently accessible even if the pod gets recycled and the IP address changes.
+
+The port that the charm's service should be reachable on is typically a config option; for example, by including a section like this in `charmcraft.yaml`:
+
+```yaml
+options:
+  config:
+    server-port:
+      type: int
+      description: the port on which to offer the service
+      default: 8000
+```
+
+This means that the charm should update the open ports in the `config-changed` handler. In the `src/charm.py` file, in the `__init__` function of your charm, set up an observer for the `config-changed` event. For example:
+
+```python
+self.framework.observe(self.on.config_changed, self._on_config_changed)
+```
+
+> See more: `[`ops.ConfigChangedEvent`](https://ops.readthedocs.io/en/latest/#ops.ConfigChangedEvent)
+
+Now, in the body of the charm definition, define the event handler, and set the port that should be open - this defaults to a TCP port, but UDP and ICMP can also be specified. For example:
+
+```python
+def _on_config_changed(self, event: ops.ConfigChangedEvent):
+    self.unit.set_ports(self.config["server-port"])
+```
+
+> Examples: [`loki-k8s` sets the open port in the charm `__init__`](https://github.com/canonical/loki-k8s-operator/blob/2e4674d64b692f82ae3dec8f7f1f0745f087827f/src/charm.py#L150), [`mysql-router` sets ports based on whether the charm is configured to be externally accessible](https://github.com/canonical/mysql-router-operator/blob/04c7b3a0ebb6279648bd3f72c5c04b0f7bad1e0c/src/machine_charm.py#L111)
+
+> See more: [`ops.Unit.set_ports`](https://ops.readthedocs.io/en/latest/#ops.Unit.set_ports)
+
+<a href="#heading--test-the-port-is-opened"><h2 id="heading--test-the-port-is-opened">Test the port is opened</h2></a>
+
+> See first: [Get started with charm testing](https://juju.is/docs/sdk/get-started-with-charm-testing)
+
+You'll want to add three levels of tests, unit, Scenario, and integration.
```

Comment:
As for the add feature section above, I'd make it more explicit that the procedure here is the same as for testing add config, even if we are nice and go on to make the recipe specific to open port as well. I'd also make the See more link down to the generic Add config doc.

Reply:
> I'd also make the See more link down to the generic Add config doc.

Unfortunately, the how-to for adding config doesn't have any details on writing tests (just a suggestion to manually check by running `juju config`) so there doesn't seem to be anywhere to link to for this.

Reply:
Tony: don't be lazy, just add the tests in that other doc and link to them here :laughing:.